### ISSUE TITLE
Variable number of points for each pointcloud during training

### ISSUE THREAD

**Issue by brbrhpt** [charlesq34/pointnet#161]
Hi, 

thanks for the nice work!

As in [#41](https://github.com/charlesq34/pointnet/issues/41), I would like to train on point clouds with very different density. 
I exchanged `tf_util.max_pool2d`  with `tf.reduce_max` as @charlesq34 suggested. However, as he also pointed out, the same number of points has to be used for each batch.

I tried to set the batch size to 1. But even when decreasing the learning rate by factor 10 or 100, the model did not converge anymore.

Is there a possibility to get the model to converge with batch size = 1?
Does it make sense, to use varying size inputs for training? (To me, it intuitively does, since during datacapture with a scanner varying density does arise)
If it does make sense, is there a way to train with varying point cloud density in tensorflow?

Sincerly,
Barbara

--
Sidenote: It seems pytorch supports variable size inputs ([https://discuss.pytorch.org/t/how-to-create-a-dataloader-with-variable-size-input/8278](https://discuss.pytorch.org/t/how-to-create-a-dataloader-with-variable-size-input/8278))

**Comment by jiachens** [comment]
Hi,

Have you resolved this?

**Comment by charlesq34** [comment]
HI @brbrhpt 

As there are batch norm layers, using batch size = 1 would cause some issues. At training time, TensorFlow does require a fixed tensor size, but you can achieve the "effect" of training with variable number of points by randomly duplicating K existing points to N points (e.g. K is varying, N is fixed as 1024)-- for pointnet, the results would be roughly the same (just a difference caused by batchnorm) as training with K points.

At test time by setting batch size as 1, you can always test with variable number of points.

**Comment by off99555** [comment]
@charlesq34 I tried training by duplicating the points and it works but not sure if it works roughly the same. I want to ask to be sure about mean/std normalization.

When training I have fixed 100 points (usually my dataset have around 40-70 points). So the extra points are padded by duplications.
When testing I want to test with 40-70 points not 100 points.

Suppose that I do standard scaling for each point cloud on the X Y Z coordinate, should I do it before or after I duplicate the points?

If I duplicate before I normalize, this will affect the coordinate values. But maybe it will make the model generalize better by not relying on specific coordinate value?
If I duplicate after I normalize, this will be like I'm just appending extra points when training. I'm not sure which approach is better. Please enlighten me.

**Comment by Maheshiitmandi** [comment]
I also want to know this.

**Comment by off99555** [comment]
@Maheshiitmandi But normalizing before duplications is also having its benefits. Suppose you have 70 points and you want to expand to 100 points, if you normalize before duplicates you will get the same prediction from the model (if you are doing classification), because most of the layers just do max pooling. And same prediction for 70 points and 100 points is what you want because the extra 30 points are just for padding.

**Comment by tkaleczyc-ats** [comment]
Hi! Does TF indeed require a fixed size input for all training samples or is this just the requirement of the same size of input within a batch? If it's the latter, maybe making artificial batches of a single point cloud permutated randomly could help?

**Comment by adosar** [comment]
> HI @brbrhpt
> 
> As there are batch norm layers, using batch size = 1 would cause some issues. At training time, TensorFlow does require a fixed tensor size, but you can achieve the "effect" of training with variable number of points by randomly duplicating K existing points to N points (e.g. K is varying, N is fixed as 1024)-- for pointnet, the results would be roughly the same (just a difference caused by batchnorm) as training with K points.
> 
> At test time by setting batch size as 1, you can always test with variable number of points.

Could we also just zero padding instead of duplication?

**Comment by adosar** [comment]
Hello @charlesq34,

> for pointnet, the results would be roughly the same (just a difference caused by batchnorm) as training with K points.

However, even with batchnorm the duplicated points would be transformed in the same way, and thanks to max pooling the output would be the same. So the only difference would be how the batch norm statistics (mean and std) and its parameters are estimated, right?
